File: Assets/src/model/items/content/ItemMushroom.cs
using System;
using System.Collections.Generic;
using UnityEngine;

[Serializable]
[ObjectInfo("mushroom", "They have such an interesting taste!")]
class ItemMushroom : Item, IStackable, IEdible {
  public ItemMushroom(int stacks) {
    this.stacks = stacks;
  }
  public int stacksMax => 100;

  private int _stacks;
  public int stacks {
    get => _stacks;
    set {
      if (value < 0) {
        throw new ArgumentException("Setting negative stack!" + this + " to " + value);
      }
      _stacks = value;
      if (_stacks == 0) {
        Destroy();
      }
    }
  }

  public void Eat(Actor a) {
    a.statuses.Add(new PumpedUpStatus(stacks));
    stacks = 0;
  }

  internal override string GetStats() => $"Eat all to make your next {stacks} attacks twice as fast.";
}

[System.Serializable]
[ObjectInfo("mushroom", "Yummy")]
class PumpedUpStatus : StackingStatus, IActionCostModifier, IActionPerformedHandler {
  public PumpedUpStatus(int stacks) : base(stacks) {}

  public ActionCosts Modify(ActionCosts input) {
    input[ActionType.ATTACK] /= 2;
    return input;
  }

  public override string Info() => $"Your next {stacks} attacks are twice as fast!";

  public void HandleActionPerformed(BaseAction final, BaseAction initial) {
    if (final.Type == ActionType.ATTACK) {
      stacks--;
    }
  }
}

File: web/src/model/items/ItemMushroom.ts
import { Item, STACKABLE_TAG, EDIBLE_TAG, type IStackable, type IEdible } from '../Item';
import type { Actor } from '../Actor';
import { PumpedUpStatus } from '../statuses/PumpedUpStatus';

/**
 * Stackable edible. Eating all stacks grants PumpedUpStatus (2x attack speed for N attacks).
 * Port of C# ItemMushroom.cs.
 */
export class ItemMushroom extends Item implements IStackable, IEdible {
  readonly [STACKABLE_TAG] = true as const;
  readonly [EDIBLE_TAG] = true as const;

  readonly stacksMax = 100;
  private _stacks: number;

  get stacks(): number {
    return this._stacks;
  }

  set stacks(value: number) {
    if (value < 0) throw new Error('Setting negative stack! ' + this + ' to ' + value);
    this._stacks = value;
    if (this._stacks === 0) {
      this.Destroy();
    }
  }

  constructor(stacks = 1) {
    super();
    this._stacks = stacks;
  }

  eat(actor: Actor): void {
    actor.statuses.add(new PumpedUpStatus(this.stacks));
    this.stacks = 0;
  }

  getStats(): string {
    return `Eat all to make your next ${this.stacks} attacks twice as fast.`;
  }
}
